### PR TITLE
core: add util plugin name parser

### PR DESCRIPTION
**Pull request by FeelyChau**

### CODE REVIEW

File: packages/core/src/types/plugins.ts [hunk 1]
```diff
@@ -10,6 +10,12 @@ interface InsertParams {
 
 export type PluginTypeI = 'dataCollect' | 'dataAccess' | 'dataProcess' | 'datasetProcess' | 'modelLoad' | 'modelDefine' | 'modelTrain' | 'modelEvaluate';
 
+// TODO(feely): remove 'fs'
```
Comment (yorkie) [inline packages/core/src/types/plugins.ts:13]: Why do we remove the fs support?

File: packages/core/src/utils/public.ts [hunk 2]
```diff
@@ -321,3 +322,23 @@ export function shuffle(array: any[]): void {
...
+  if (path.isAbsolute(name)) {
+    protocol = 'fs';
+  } else if (/^git(\+ssh|\+https|\+http)?:$/.test(urlObject.protocol)) {
+    protocol = 'git';
+  } else if ([ 'https:', 'http:' ].indexOf(urlObject.protocol) !== -1) {
+    protocol = 'tarball';
+  } else if (name[0] !== '.') {
+    protocol = 'npm';
+  } else {
+    protocol = undefined;
```
Comment (yorkie) [inline packages/core/src/utils/public.ts:341]: Throw an error?

File: packages/core/src/utils/public.ts [hunk 3]
```diff
@@ -321,3 +322,23 @@ export function shuffle(array: any[]): void {
...
+ * parse plugin name, return the plugin protocol and url object
+ */
+export function parsePluginName(name: string): { protocol: PluginProtocol; urlObject: url.UrlWithStringQuery } {
+  const urlObject = url.parse(name);
+  let protocol: PluginProtocol;
+  if (path.isAbsolute(name)) {
+    protocol = 'fs';
+  } else if (/^git(\+ssh|\+https|\+http)?:$/.test(urlObject.protocol)) {
+    protocol = 'git';
+  } else if ([ 'https:', 'http:' ].indexOf(urlObject.protocol) !== -1) {
```
Comment (yorkie) [inline packages/core/src/utils/public.ts:336]: This doesn't mean tarbar, just for http/https url.

File: packages/core/src/utils/public.ts [hunk 4]
```diff
@@ -321,3 +322,23 @@ export function shuffle(array: any[]): void {
 export function generateId(): string {
   return nanoid();
 }
+
+/**
+ * parse plugin name, return the plugin protocol and url object
+ */
+export function parsePluginName(name: string): { protocol: PluginProtocol; urlObject: url.UrlWithStringQuery } {
```
Comment (yorkie) [inline packages/core/src/utils/public.ts:329]: public_utils is mostly for the plugin developer, so IMO this function should be another utils module.

File: packages/costa/src/runtime.ts [hunk 5]
```diff
@@ -247,22 +246,22 @@ export class CostaRuntime {
   async fetch(name: string): Promise<PluginPackage> {
     let pkg: PluginPackage;
     const source = this.getSource(name);
-    if (source.from === 'npm') {
+    if (source.protocol === 'npm') {
```
Comment (yorkie) [inline packages/costa/src/runtime.ts:249]: May I ask you not to change the field name?

File: packages/core/src/utils/public.ts (hunk 3 above)
Comment (FeelyChau) [reply packages/core/src/utils/public.ts:336]: For plugin, it's tarball.

File: packages/core/src/types/plugins.ts (hunk 1 above)
Comment (FeelyChau) [reply packages/core/src/types/plugins.ts:13]: OK, the `fs` protocol is for tarball-stream which upload from client, rename to `tarball-stream`?

File: packages/costa/src/runtime.ts (hunk 5 above)
Comment (FeelyChau) [reply packages/costa/src/runtime.ts:249]: I think the field name is not very accurate. IMO, `protocol` + `uri` equal `from`, `protocol` is a little bit more accurate.

File: packages/core/src/utils/public.ts (hunk 4 above)
Comment (FeelyChau) [reply packages/core/src/utils/public.ts:329]: OK, should we create a new utils package?

File: packages/core/src/types/plugins.ts (hunk 1 above)
Comment (FeelyChau) [reply packages/core/src/types/plugins.ts:13]: Protocol is the concept that needs to be passed to the user,`fs` is the implementation, should it be changed to a more understandable name, like `tarball-stream` or `local-tarball`?

File: packages/core/src/utils/public.ts (hunk 2 above)
Comment (FeelyChau) [reply packages/core/src/utils/public.ts]: Updated.

File: packages/core/src/utils/public.ts (hunk 3 above)
Comment (yorkie) [reply packages/core/src/utils/public.ts:336]: I think we need to verify if it ends with .tgz or .gz, just checking for protocol is not enough in this case.

File: packages/core/src/types/plugins.ts (hunk 1 above)
Comment (yorkie) [reply packages/core/src/types/plugins.ts]: Ok, you should update what you commented before in the TODO.

File: packages/core/src/utils/public.ts (hunk 4 above)
Comment (yorkie) [reply packages/core/src/utils/public.ts:329]: No new package, they could be put in core/src/utils.

File: packages/costa/src/runtime.ts (hunk 5 above)
Comment (yorkie) [reply packages/costa/src/runtime.ts:249]: This PR is not to change it, and it makes some unexpected changes.

File: packages/core/src/utils/public.ts (hunk 4 above)
Comment (FeelyChau) [reply packages/core/src/utils/public.ts:329]: OK.

File: packages/core/src/utils/public.ts (hunk 3 above)
Comment (FeelyChau) [reply packages/core/src/utils/public.ts:336]: OK, it's better to check the path.

File: packages/core/src/types/plugins.ts [hunk 6]
```diff
@@ -10,6 +10,11 @@ interface InsertParams {
 
 export type PluginTypeI = 'dataCollect' | 'dataAccess' | 'dataProcess' | 'datasetProcess' | 'modelLoad' | 'modelDefine' | 'modelTrain' | 'modelEvaluate';
 
+/**
+ * The plugin installation protocol
+ */
+export type PluginProtocol = 'fs' | 'npm' | 'git' | 'tarball';
```
Comment (yorkie) [inline packages/core/src/types/plugins.ts:16]:
```suggestion
export type PluginSourceProtocol = 'fs' | 'npm' | 'git' | 'tarball';
```

File: packages/core/src/index.ts [hunk 7]
```diff
@@ -60,6 +61,10 @@ export {
   generateId
 } from './utils/public';
 
+export {
+  parsePluginName
+} from './utils/utils';
```
Comment (yorkie) [inline packages/core/src/index.ts:66]: Hmm, `utils` is duplicated with the folder, suggestion: `internals` or `common`.

File: packages/costa/src/runtime.ts [hunk 8]
```diff
@@ -477,32 +476,24 @@ export class CostaRuntime {
...
+    } else if (protocol === 'npm') {
       src.schema = this.getNameSchema(name);
-      src.from = 'npm';
       let { npmRegistryPrefix } = this.options;
       if (npmRegistryPrefix.slice(-1) === '/') {
         npmRegistryPrefix = npmRegistryPrefix.slice(0, -1);
       }
       src.uri = `${npmRegistryPrefix}/${src.schema.packageName}`;
-    } else {
-      throw new TypeError(`Unsupported resolving plugin name: ${name}`);
```
Comment (yorkie) [inline packages/costa/src/runtime.ts:505]: I think we still need the throw.

**Comment by yorkie** [comment]
LGTM except for the filename, @FeelyChau could you please rename it to a proper filename?

**Comment by FeelyChau** [comment]
> LGTM except for the filename, @FeelyChau could you please rename it to a proper filename?

`common.ts`?
Or change the tree to:
```shell
src
|--public.ts
|--utils.ts
```

**Review by yorkie — APPROVED** [review]
LGTM if CI is green

**Pull request merged**